### PR TITLE
Port to cryptography (fixes #16)

### LINKED ISSUE #16: Port to cryptography and drop M2Crypto dependency
M2Crypto is a beast to install and requires unofficial patches on many Linux distros now. I think we can port encrypted cookies to use [cryptography](https://cryptography.io/en/latest/) instead but I haven't looked at it in detail.

### PR DESCRIPTION
**Pull request by kumar303**
The cryptography library is easier to deal with and better maintained than M2Crypto. There are a few other changes in here which are listed in the changelog.

### CODE REVIEW

**Comment by frankoid** [comment]
Since it is (not trivial to install cryptography on OS X)[http://chriskief.com/2014/03/25/installing-cryptography-via-pip-with-macports-or-homebrew/] I would rather not make it the only supported backend. On my machine (OS X 10.10, openssl installed via MacPorts) tox fails with the following error:
```
  File "/Users/francis/Code/Bright/python/django-encrypted-cookie-session/.tox/py26-django16/lib/python2.6/site-packages/cffi/vengine_cpy.py", line 155, in load_library
    raise ffiplatform.VerificationError(error)
VerificationError: importing '/Users/francis/Code/Bright/python/django-encrypted-cookie-session/.tox/py26-django16/lib/python2.6/site-packages/cryptography/_Cryptography_cffi_966afe82xd5c405b7.so': dlopen(/Users/francis/Code/Bright/python/django-encrypted-cookie-session/.tox/py26-django16/lib/python2.6/site-packages/cryptography/_Cryptography_cffi_966afe82xd5c405b7.so, 2): Symbol not found: _CRYPTO_malloc_debug_init
  Referenced from: /Users/francis/Code/Bright/python/django-encrypted-cookie-session/.tox/py26-django16/lib/python2.6/site-packages/cryptography/_Cryptography_cffi_966afe82xd5c405b7.so
  Expected in: flat namespace
 in /Users/francis/Code/Bright/python/django-encrypted-cookie-session/.tox/py26-django16/lib/python2.6/site-packages/cryptography/_Cryptography_cffi_966afe82xd5c405b7.so
```

Would you be able to make the backends switchable (as used to be the case for M2Crypto and pycrypto)?

File: README.md [hunk 1]
```diff
@@ -147,15 +151,31 @@ need 2.6 as well to run all environments. Run the tests like this:
...
 Changelog
 =========
 
+3.0.0
+-----
+
+* Dropped support for [M2Crypto](https://pypi.python.org/pypi/M2Crypto)
+  in favor of [cryptography](https://cryptography.io/en/latest/) for
+  better platform support.
+* **BREAKING CHANGE**: Old values from the `ENCRYPTED_COOKIE_KEY` setting
```
Comment (frankoid) [inline README.md]: Are these breaking changes due to the switch to cryptography or something else? In other words, if we supported both cryptography and M2Crypto would users who kept using M2Crypto need to change their key settings or not?

Comment (kumar303) [inline README.md]: yeah, it was only due to cryptography since you have to use a base 64 encoded 32 byte key.

**Comment by kumar303** [comment]
Oh yeah, good idea. I can try to take a stab at making the backend switchable.

**Comment by kumar303** [comment]
> Since it is (not trivial to install cryptography on OS X)

The blog you linked to is pretty old and the bug they link to is closed. Did you try the official install docs? https://cryptography.io/en/latest/installation/#using-your-own-openssl-on-os-x I've only been installing it on Linux via Docker which was pretty straightforward. The install docs are very thorough for each platform.

**Comment by frankoid** [comment]
The first thing I tried was just running `tox`, which failed with the `Symbol not found: _CRYPTO_malloc_debug_init` error I mentioned earlier.
If I run it as `env ARCHFLAGS="-arch x86_64" LDFLAGS="-L/opt/local/lib" CFLAGS="-I/opt/local/include" tox` then all the tests do succeed.

Thanks for the link to the official install docs, I hadn't read them before. They recommend pretty much the same thing as the old blog article though (setting environment variables so that the MacPorts or Homebrew OpenSSL headers and libs are used).

In an ideal world such environment variables wouldn't be necessary - it'd be nice to be able to just put:
```
django-encrypted-cookie-session==3.0.0
cryptography==x.y.z
```
in requirements.txt and do a plain `pip install -r requirements.txt` and have it work properly, rather than having to set environment variables for cryptography's sake.

In short, switching from M2Crypto to cryptography seems to reduce installation hassle on Linux but increase installation hassle on OS X which is why I'm questioning it.

**Comment by kumar303** [comment]
m2crypto has been around a while (it was unmaintained but I think it got a new maintainer). It's heavyweight but people generally know how to tame it. Cryptography is newer but seems to have strong support from the Python community and a lot of momentum currently. I'll see how hard it is to create a switchable backend so that people can choose the one they want. 

Also, I'll add a note in the developer docs for the environment var fiddling that may be necessary when running tox.

**Comment by frankoid** [comment]
Thanks. Interestingly, M2Crypto manages to find and link to the MacPorts openssl without the need for any environment variables, so in theory it should be possible to make cryptography do so too:
```
$ otool -L .tox/py27-django14-m2c/lib/python2.7/site-packages/M2Crypto/__m2crypto.so
.tox/py27-django14-m2c/lib/python2.7/site-packages/M2Crypto/__m2crypto.so:
	/opt/local/lib/libssl.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/opt/local/lib/libcrypto.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)
```

**Comment by kumar303** [comment]
> The first thing I tried was just running tox, which failed with the Symbol not found: _CRYPTO_malloc_debug_init error I mentioned earlier.

Just wanted to mention that I never saw this error on Mac 10.10 running the tests in tox. How odd. In other words, I did not need to set any environment variables to get tox to build the virtualenv with cryptography in it.

**Comment by frankoid** [comment]
> Just wanted to mention that I never saw this error on Mac 10.10 running the tests in tox. How odd. In other words, I did not need to set any environment variables to get tox to build the virtualenv with cryptography in it.

I wonder whether I got the error because I have OpenSSL installed in MacPorts. Do you use MacPorts or Homebrew, and if so do you have OpenSSL installed in either of them?

**Comment by kumar303** [comment]
Right, that could be why. Perhaps if you had the stock Apple openssl then it would have worked without an error. I use homebrew and it's very likely that I already had their openssl installed so I can't say for sure what happens with Apple's stock lib.

**Comment by kumar303** [comment]
@frankoid I made the backends configurable. Can you take another look?

**Comment by frankoid** [comment]
Ironically I can't get tox to pass with the new optional M2Crypto support.

The problem is that the same environment variables that I need in order to get cryptography to install prevent M2Crypto from installing successfully (I can install M2Crypto OK when they're not defined). (Here are the details)[https://gist.github.com/frankoid/b1511075df88d3b1ee7f].

Given this, I think that it might be best to just drop support for M2Crypto as per your original PR. Is it OK with you if I just merge up to the commit before the one where you introduced the optional M2Crypto support? I'm really sorry for making you go to the extra effort, I'm just trying to avoid succumbing to the sunk cost fallacy.

**Comment by kumar303** [comment]
No worries at all. Installation and usage is a lot simpler with only one backend. What I did was revert the backend adapter commit so that for some crazy reason we need it in the future, it will be there in the history. This should be ready to merge as is, thanks for the reviews!

File: encrypted_cookies/crypto.py [hunk 2]
```diff
@@ -1,28 +1,34 @@
...
-    return key
+from  cryptography import fernet
 
 
 def encrypt(plaintext_bytes):
-    key = encrypted_cookie_key()
-    if not key:
-        raise ValueError('The ENCRYPTED_COOKIE_KEY and SECRET_KEY settings must not both be empty.')
+    """
+    Returns an ecnrypted version of plaintext_bytes.
```
Comment (frankoid) [inline encrypted_cookies/crypto.py]: s/ecnrypted/encrypted/

Comment (frankoid) [inline encrypted_cookies/crypto.py]: I've fixed this

**Comment by kumar303** [comment]
Great, thanks! You want to push it up to PyPI? I upgraded an app and it seemed to work but we could always follow up with a point release if it needs fixes.

**Comment by frankoid** [comment]
Pushed to PyPI :) Thanks for this change, the new minimal crypto.py is great!
